crypto/kdf/kdf_util.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2018, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdarg.h>
+#include "internal/cryptlib.h"
+#include "internal/evp_int.h"
+#include <openssl/kdf.h>
+#include "internal/numbers.h"
+#include <openssl/evp.h>
+#include "kdf_local.h"
+
+int call_ctrl(int (*ctrl)(EVP_KDF_IMPL *impl, int cmd, va_list args),
+              EVP_KDF_IMPL *impl, int cmd, ...)
+{
+    int ret;
+    va_list args;
+
+    va_start(args, cmd);
+    ret = ctrl(impl, cmd, args);
+    va_end(args);
+
+    return ret;
+}
+
+/* Utility functions to send a string or hex string to a ctrl */
+
+int kdf_str2ctrl(EVP_KDF_IMPL *impl,
+                 int (*ctrl)(EVP_KDF_IMPL *impl, int cmd, va_list args),
+                 int cmd, const char *str)
+{
+    return call_ctrl(ctrl, impl, cmd, (const unsigned char *)str, strlen(str));
+}
+
+int kdf_hex2ctrl(EVP_KDF_IMPL *impl,
+                 int (*ctrl)(EVP_KDF_IMPL *impl, int cmd, va_list args),
+                 int cmd, const char *hex)
+{
+    unsigned char *bin;
+    long binlen;
+    int ret = -1;
+
+    bin = OPENSSL_hexstr2buf(hex, &binlen);
+    if (bin == NULL)
+        return 0;
+
+    if (binlen <= INT_MAX)
+        ret = call_ctrl(ctrl, impl, cmd, bin, (size_t)binlen);
+    OPENSSL_free(bin);
+    return ret;
+}
+
+/* Pass a message digest to a ctrl */
+int kdf_md2ctrl(EVP_KDF_IMPL *impl,
+                int (*ctrl)(EVP_KDF_IMPL *impl, int cmd, va_list args),
+                int cmd, const char *md_name)
+{
+    const EVP_MD *md;
+
+    if (md_name == NULL || (md = EVP_get_digestbyname(md_name)) == NULL) {
+        KDFerr(KDF_F_KDF_MD2CTRL, KDF_R_INVALID_DIGEST);
+        return 0;
+    }
+    return call_ctrl(ctrl, impl, cmd, md);
+}
```

Comment:
I wouldn't mind at all if these got generalised into the public API.  Suggested names: `EVP_vstr2ctrl`, `EVP_vhex2ctrl`, `EVP_vmd2ctrl`.

Reply:
Hmmm. I don't think we need them yet, at least not for KDF. They wouldn't be needed to help with implementing KDF algorithms as there is currently no way to do that.

I have tried to avoid adding any brand new legacy stuff, such as ENGINE support, since we don't know how the proposed provider model will pan out. There is currently no way to set methods on an `EVP_KDF_METHOD`, for example, and `EVP_KDF_METHOD` isn't even defined in `include/openssl/evp.h` (or any public header). That's mainly because if/when it does become public it will probably be called `EVP_KDF`, analogous to the new `EVP_MAC`.

Reply:
I'm not sure how ENGINE came into play...  I was rather thinking of function that could join the already implemented `EVP_str2ctrl` and `EVP_hex2ctrl`, the only difference being that these are more elegantly suited for use with variadic style control functions.

Reply:
Yeah, okay. I was just trying to guess the use case for those being in the public API.

Reply:
@levitte , why did you add `EVP_str2ctrl` and `EVP_hex2ctrl` to the public API?

Reply:
Because I saw a general usefulness...  I don't think we're getting rid of control functions any time soon, regardless of anything else, so I believe these kinds of utility functions have their place there.